tensorflow_lite_support/cc/test/task/processor/image_preprocessor_test.cc

```diff
@@ -0,0 +1,147 @@
+using ::tflite::task::vision::FrameBuffer;
+using ::tflite::task::vision::ImageData;
+
+constexpr char kTestDataDirectory[] =
+    "tensorflow_lite_support/cc/test/testdata/task/vision/";
+
+constexpr char kDilatedConvolutionModelWithMetaData[] = "dilated_conv.tflite";
+
+StatusOr<ImageData> LoadImage(std::string image_name) {
+  return DecodeImageFromFile(
+      JoinPath("./" /*test src dir*/, kTestDataDirectory, image_name));
+}
+
+class DynamicInputTest : public tflite_shims::testing::Test {
+ public:
+  void SetUp() {
+    engine_ = absl::make_unique<TfLiteEngine>();
+    engine_->BuildModelFromFile(JoinPath("./", kTestDataDirectory,
+                                         kDilatedConvolutionModelWithMetaData));
+    engine_->InitInterpreter();
+
+    SUPPORT_ASSERT_OK_AND_ASSIGN(preprocessor_,
+                                 ImagePreprocessor::Create(engine_.get(), {0}));
+  }
+
+ protected:
+  std::unique_ptr<ImagePreprocessor> preprocessor_ = nullptr;
+  std::unique_ptr<TfLiteEngine> engine_ = nullptr;
+};
+
+// See if input tensor dims signature for height and width is -1
+// because it is so in the model.
+TEST_F(DynamicInputTest, InputHeightAndWidthMutable) {
+  const TfLiteIntArray *input_dims_signature =
+      engine_->GetInputs()[0]->dims_signature;
+  EXPECT_EQ(input_dims_signature->data[1], -1);
+  EXPECT_EQ(input_dims_signature->data[2], -1);
+}
+
+// See if output tensor has been re-dimmed as per the input
+// tensor. Expected shape: (1, input_height, input_width, 16).
+TEST_F(DynamicInputTest, OutputDimensionCheck) {
+  SUPPORT_ASSERT_OK_AND_ASSIGN(ImageData image, LoadImage("burger.jpg"));
+  std::unique_ptr<FrameBuffer> image_frame_buffer = CreateFromRgbRawBuffer(
+      image.pixel_data, FrameBuffer::Dimension{image.width, image.height});
+
+  preprocessor_->Preprocess(*image_frame_buffer);
+  absl::Status status = engine_->interpreter_wrapper()->InvokeWithoutFallback();
+  EXPECT_TRUE(status.ok());
+  EXPECT_EQ(engine_->GetOutputs()[0]->dims->data[0], 1);
+  EXPECT_EQ(engine_->GetOutputs()[0]->dims->data[1],
+            engine_->GetInputs()[0]->dims->data[1]);
+  EXPECT_EQ(engine_->GetOutputs()[0]->dims->data[2],
+            engine_->GetInputs()[0]->dims->data[2]);
+  EXPECT_EQ(engine_->GetOutputs()[0]->dims->data[3], 16);
+}
+
+// Compare pre-processed input with an already pre-processed
+// golden image.
+TEST_F(DynamicInputTest, GoldenImageComparison) {
+  SUPPORT_ASSERT_OK_AND_ASSIGN(ImageData image, LoadImage("burger.jpg"));
+  std::unique_ptr<FrameBuffer> image_frame_buffer = CreateFromRgbRawBuffer(
+      image.pixel_data, FrameBuffer::Dimension{image.width, image.height});
+
+  preprocessor_->Preprocess(*image_frame_buffer);
+
+  // Check the processed input image.
+  float *processed_input_data =
+      tflite::task::core::AssertAndReturnTypedTensor<float>(
+          engine_->GetInputs()[0]);
+
+  bool is_equal = true;
+  float epsilon = 0.1f;
+  std::string file_path =
```

Comment:
How about compare `processed_input_data` with `image`? Theoretically, `image` is not changed in Preprocess(), i.e. no cropping or resizing, except that the values converted from uint8 to float.

Reply:
I suppose you mean `image` is the `golden image`? The problem here is, `Preprocess()` normalizes the image, which is why `processed_input_data` is a `float*`. Now, to compare, we need the "normalized" image each value would be float in range (-1, 1). 

`float` can't be stored in `jpg` so I'm storing them in `bin` file.

Reply:
Used [this](https://gist.github.com/jonpsy/0b6b9069a3b6cb0d024c7163b30abaa6) script to normalize and store `burger.jpg`

Reply:
Please see the other comment about bypassing the normalization. Testing against the original image is more robust and easy to debug than comparing with the bin file.